PHPCompatibility/Sniffs/FunctionDeclarations/RemovedImplicitlyNullableParamSniff.php

```diff
@@ -0,0 +1,147 @@
+        $parameters = FunctionDeclarations::getParameters($phpcsFile, $stackPtr);
+        if (empty($parameters)) {
+            return;
+        }
+
+        $tokens = $phpcsFile->getTokens();
+
+        // Parenthesis closer will be defined, otherwise we'd have received an empty parameters array.
+        $closeParens = $tokens[$stackPtr]['parenthesis_closer'];
+
+        foreach ($parameters as $key => $param) {
+            if (isset($param['property_visibility'])) {
+                /*
+                 * Implicitly nullable parameters were never allowed for promoted properties
+                 * and always resulted in a fatal error. Ignore as this is outside the scope of this sniff.
+                 */
+                continue;
+            }
+
+            if (isset($param['default']) === false
+                || $param['type_hint'] === ''
+                || $param['type_hint_token'] === false
+            ) {
+                // If there is no default value or no type hint, there is no issue.
+                continue;
+            }
+
+            if ($param['nullable_type'] === true
+                || $param['type_hint'] === 'null'
+                || $param['type_hint'] === 'mixed'
+                || $phpcsFile->findNext(\T_NULL, $param['type_hint_token'], ($param['type_hint_end_token'] + 1)) !== false
+            ) {
+                // Type is nullable, no issue.
+                continue;
+            }
+
+            /*
+             * Non-nullable type, now check if the default value is `null`.
+             *
+             * We can ignore `null` when it is part of a constant expression or in new in initializers,
+             * as PHP does not regard parameters with that kind of default value as nullable and
+             * would throw a fatal when the function is called. This behaviour is not changed by the
+             * PHP 8.4 deprecation of implicitly nullable parameters, so those parameters should be
+             * ignored by this sniff.
+             * Also see: https://github.com/php/php-src/issues/13752
+             */
+
+            // Determine the end of the parameter.
+            $paramEnd   = ($param['comma_token'] === false) ? $closeParens : $param['comma_token'];
+            $hasNull    = $phpcsFile->findNext(\T_NULL, $param['default_token'], $paramEnd);
+            $hasNonNull = $phpcsFile->findNext($this->allowedInDefault, $param['default_token'], $paramEnd, true);
+            if ($hasNull === false || $hasNonNull !== false) {
+                // No null default value, we're okay.
+                continue;
+            }
+
+            $error = 'Implicitly marking a parameter as nullable is deprecated since PHP 8.4. Update the type to be explicitly nullable instead. Found implicitly nullable parameter: %s.';
+            $code  = 'Deprecated';
+            $data  = [$param['name']];
+
+            $phpcsFile->addWarning($error, $param['token'], $code, $data);
```

Comment:
I feel that, similar to the CurlyBraceArrayAccess sniff, this is something that can be quite time-consuming to fix (due to the sheer number of cases) and is trivial to fix automatically.

Here's what you said on that commit (e89af9203a686b1c08057f6480190e96517479e3)
>  For larger codebases, this issue can be quite time-consuming to fix, while fixing this automatically is trivial.

If this sniff does not include an auto-fix, then it'll probably get forked or copied to other sniffs to do so.

```php

            if ($fix === true) {
                if (strpos($param['type_hint'], '&') !== false) {
                    $hint = "null|(" . $param['type_hint'] . ")";
                } elseif (strpos($param['type_hint'], '|') !== false) {
                    $hint = "null|" . $param['type_hint'];
                } else {
                    $hint = "?" . $param['type_hint'];
                }
                $phpcsFile->fixer->beginChangeset();
                $phpcsFile->fixer->replaceToken($param['type_hint_token'], $hint);
                $phpcsFile->fixer->endChangeset();
            }
```

Reply:
As you saw in https://github.com/PHPCompatibility/PHPCompatibility/pull/1689#issuecomment-2000963275, this sniff will not get an auto-fixer for good reason.

The `CurlyBraceArrayAccess` issue only has one fix and that fix is cross-version compatible. That is not the case for this issue, where there really are many more choices, which means that it is not for PHPCompatibility to decide how it should be fixed.